### ISSUE TITLE
Unexpected @Vector() behavior with i2 items

### ISSUE THREAD

**Issue by voroskoi** [ziglang/zig#11611]
### Zig Version

0.10.0-dev.2131+9416b4d99

### Steps to Reproduce

```zig
const std = @import("std");

const dirs = [_]@Vector(2, i2){
    @Vector(2, i2){ -1, -1 },
    @Vector(2, i2){ -1, 1 },
    @Vector(2, i2){ 1, -1 },
    @Vector(2, i2){ 1, 1 },
};

test "Vector" {
    // const pos: @Vector(2, isize) = .{ 1, 1 };
    // for (dirs) |d| {
    //     const diff = pos + d;
    //     std.debug.print("{}\n", .{diff});
    // }
    const pos2: @Vector(2, usize) = .{ 1, 1 };
    for (dirs) |d| {
        const diff: @Vector(2, isize) = .{
            @intCast(isize, pos2[0]) + d[0],
            @intCast(isize, pos2[1]) + d[1],
        };
        std.debug.print("{}\n", .{diff});
    }
}
```

### Expected Behavior

Running `zig test` gives:

```
Test [1/1] test "Vector"... { 0, 0 }
{ 0, 2 }
{ 2, 0 }
{ 2, 2 }
```

Related questions:
1.  `-fno-stage1` does not compile if `dirs` is defined as tuples. Is this intentional?
2. The commented part gives `./src/main.zig:13:26: error: incompatible types: '@Vector(2, isize)' and '@Vector(2, i2)'`. It could be auto coerced, right?
3. Is there a plan to add `@Vector` support for `@intCast`, so You do not have to cast each dimension?

### Actual Behavior

Running `zig test` gives:

```
Test [1/1] test "Vector"... { 0, 1 }
{ 0, 1 }
{ 2, 1 }
{ 2, 1 }
```

Notice that the second value remains `1`. I have checked with 3 dimensions too, only the first changes. Using `-fno-stage1` gives the same result.

However this error goes away as soon as I use `@Vector(2, isize)` when declaring `dirs`.

**Comment by ifreund** [comment]
`@intCast()` already supports vectors as of https://github.com/ziglang/zig/pull/6744... looks like that feature could use some documentation.

**Comment by wsengir** [comment]
This can be reduced into just runtime indexing of `@Vector`s of exotic signed ints in seemingly only debug builds:
```zig
const std = @import("std");

pub fn main() void {
    func(0);
}

const dirs = [_]@Vector(2, i2){
    @Vector(2, i2){ -1, -1 },
};

fn func(i: usize) void {
    {
        var right_d = dirs[0];
        std.debug.print("{}\n", .{right_d});
    }
    {
        var wrong_d = dirs[i];
        std.debug.print("{}\n", .{wrong_d});
    }
    {
        for (dirs) |wrong_d| {
            std.debug.print("{}\n", .{wrong_d});
        }
    }
}
```
prints
```
{ -1, -1 }
{ -1, 0 }
{ -1, 0 }
```

**Comment by Vexu** [comment]
Fixed by #13959
